src/core/raster/qgsrasterattributetable.h

```diff
@@ -0,0 +1,200 @@
+      QString name;
+      FieldUsage usage;
+      QVariant::Type type;
+    };
+
+    /**
+     * Returns the RAT type.
+     */
+    const RatType &type() const;
+
+    /**
+     * Sets the RAT \a type
+     */
+    void setType( const RatType &newType );
+
+    /**
+     * Returns TRUE if the RAT has RGB information.
+     */
+    bool hasColor();
+
+    /**
+     * Returns the RAT fields.
+     */
+    QList<QgsRasterAttributeTable::Field> fields() const;
+
+    /**
+     * Returns the RAT fields as QgsFields.
+     */
+    QgsFields qgisFields() const;
+
+    /**
+     * Returns the RAT rows as a list of QgsFeature.
+     */
+    QgsFeatureList qgisFeatures( ) const;
+
+    /**
+     * Returns TRUE if the RAT was modified from its last reading from the storage.
+     */
+    bool isDirty() const;
+
+    /**
+     * Sets the RAT dirty state to \a isDirty;
+     */
+    void setIsDirty( bool isDirty );
+
+    /**
+     * Returns TRUE if the RAT is valid.
+     */
+    bool isValid() const;
+
+    /**
+     * Inserts a new \a field at \a position and returns TRUE on success.
+     */
+    bool insertField( const QgsRasterAttributeTable::Field &field, int position = 0 );
+
+    /**
+     * Creates a new field from \a name, \a usage and \a type and inserts it at \a position, returns TRUE on success.
+     */
+    bool insertField( const QString &name, QgsRasterAttributeTable::FieldUsage usage, QVariant::Type type, int position = 0 );
+
+    /**
+     * Creates a new field from \a name, \a usage and \a type and appends it to the fields, returns TRUE on success.
+     */
+    bool appendField( const QString &name, QgsRasterAttributeTable::FieldUsage usage, QVariant::Type type );
+
+    /**
+     * Appends a new \a field and returns TRUE on success.
+     */
+    bool appendField( const QgsRasterAttributeTable::Field &field );
+
+    /**
+     * Removes the field with \a name, returns TRUE on success.
+     */
+    bool removeField( const QString &name );
+
+    /**
+     * Inserts a row of \a data in the RAT at \a position, returns TRUE on success.
+     */
+    bool insertRow( const QVariantList data, int position = 0 );
+
+    /**
+     * Appends a row of \a data to the RAT, returns TRUE on success.
+     */
+    bool appendRow( const QVariantList data );
+
+    /**
+     * Writes the RAT to a DBF file specified by \a path, optionally reporting any error in \a errorMessage, returns TRUE on success.
+     */
+    bool writeToFile( const QString &path, QString *errorMessage SIP_OUT = nullptr );
+
+    /**
+     * Reads the RAT from a DBF file specified by \a path, optionally reporting any error in \a errorMessage, returns TRUE on success.
+     */
+    bool readFromFile( const QString &path, QString *errorMessage SIP_OUT = nullptr );
+
+    /**
+     * Returns the RAT rows.
+     */
+    const QList<QList<QVariant>> &data() const;
```

Comment:
Return by value -- there's no benefit in returning qt implicitly shared classes by ref

Reply:
Yeah, I tend to use the QtCreator refactoring tools to create getters and setters and this is what they do, I'll fix that.